sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionKeyRange.java

```diff
@@ -76,20 +78,28 @@ public String getMinInclusive() {
         return super.getString("minInclusive");
     }
 
-    public void setMinInclusive(String minInclusive) {
+    private void setMinInclusive(String minInclusive) {
         BridgeInternal.setProperty(this, "minInclusive", minInclusive);
     }
 
     public String getMaxExclusive() {
         return super.getString("maxExclusive");
     }
 
-    public void setMaxExclusive(String maxExclusive) {
+    private void setMaxExclusive(String maxExclusive) {
         BridgeInternal.setProperty(this, "maxExclusive", maxExclusive);
     }
 
     public Range<String> toRange() {
-        return new Range<String>(this.getMinInclusive(), this.getMaxExclusive(), true, false);
+        if (this.range == null) {
+            synchronized (this) {
```

Comment:
Loud thinking  - If toRange () method return same value for a PartitionKeyRange object , I am seeing a trade off between locking (making sure only one thread create object ) vs time saving if we dont do synchronized and letting two thread create object if they simultaneously comes in this block

Reply:
as per Happen-Before relationship in java. To allow one thread sees the effect of the other, either this field has to be volatile or we should do synchronized lock.

as synchrnozied is only used on set assignment it doesn't have any cost on the get, however there is always a cost associated with get in volatile. synchronized doesn't add any cost on the get which is the intention.

Reply:
As i said in this particular scenario as both thread end up creating same value we can deviate from Happens-Before , how ever as you mentioned synchronized  doesn't add any cost then this is good . All set